firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java

```diff
@@ -1088,4 +1094,40 @@ public void testHandlesSetMutationThenTransformMutationThenRemoteEventThenTransf
     assertChanged(doc("foo/bar", 4, map("sum", 1339), Document.DocumentState.COMMITTED_MUTATIONS));
     assertContains(doc("foo/bar", 4, map("sum", 1339), Document.DocumentState.COMMITTED_MUTATIONS));
   }
+
+  @Test
+  public void testHandlesMergeMutationWithTransformThenRemoteEvent() {
+    Query query = Query.atPath(ResourcePath.fromString("foo"));
+    allocateQuery(query);
+    assertTargetId(2);
+
+    writeMutations(
+        asList(
+            patchMutation("foo/bar", map(), Collections.emptyList()),
+            transformMutation("foo/bar", map("sum", FieldValue.numericAdd(1)))));
+    assertChanged(doc("foo/bar", 0, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+    assertContains(doc("foo/bar", 0, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+
+    applyRemoteEvent(addedRemoteEvent(doc("foo/bar", 1, map("sum", 1337)), asList(2), emptyList()));
+    assertChanged(doc("foo/bar", 1, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+    assertContains(doc("foo/bar", 1, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+  }
+
+  @Test
+  public void testHandlesPatchMutationWithTransformThenRemoteEvent() {
+    Query query = Query.atPath(ResourcePath.fromString("foo"));
+    allocateQuery(query);
+    assertTargetId(2);
+
+    writeMutations(
+        asList(
+            patchMutation("foo/bar", map()),
+            transformMutation("foo/bar", map("sum", FieldValue.numericAdd(1)))));
+    assertChanged(deletedDoc("foo/bar", 0));
+    assertNotContains("foo/bar");
+
+    applyRemoteEvent(addedRemoteEvent(doc("foo/bar", 1, map("sum", 1337)), asList(2), emptyList()));
+    assertChanged(doc("foo/bar", 1, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
```

Comment:
FWIW, I don't think this is actually ideal behavior.  Since we won't have raised an initial event with sum=1, it would be better to apply the increment to the actual remote value here and raise sum=1338.

I think ideally we'd somehow implement this so that the _first_ time we apply the Transform and it generates a document (because we had a base document to act on), we'd freeze the base state values and use them from then on until the mutation is acknowledged.  But the current implementation doesn't lend itself to that behavior, so I think we probably want to just live with this as an edge case.

Perhaps add a comment of some kind though so if we ever rework this implementation (e.g. during the overlay work) and look at this test, we'll know that this behavior isn't necessarily ideal and we might want to tweak it.

Reply:
I added a comment.

Reply:
Oh... nice.